The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5113404288/jobs/9192565043) [(plain)](https://github.com/rust-lang/rust/commit/bd587fda061a6c6d22a11523530b6f77ffae91e0/checks/13838238632/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling core v0.0.0 (/checkout/library/core)
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
    --> /checkout/library/core/src/mem/mod.rs:1323:5
     |
1322 | pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
     | ------------------- in this expansion of `offset_of!`
1323 |     builtin # offset_of($Container, $($fields).+)
     |
    ::: library/core/tests/mem.rs:461:16
     |
     |
461  |     assert_eq!(offset_of!(Alpha, x), 0);
     |
     = help: the trait `Sized` is not implemented for `[u8]`

error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
    --> /checkout/library/core/src/mem/mod.rs:1323:5
     |
1322 | pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
     | ------------------- in this expansion of `offset_of!`
1323 |     builtin # offset_of($Container, $($fields).+)
     |
    ::: library/core/tests/mem.rs:462:16
     |
     |
462  |     assert_eq!(offset_of!(Alpha, y), 2);
     |
     = help: the trait `Sized` is not implemented for `[u8]`


error[E0277]: the size for values of type `(dyn mem::offset_of_dst::Trait + 'static)` cannot be known at compilation time
     |
     |
1322 | pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
     | ------------------- in this expansion of `offset_of!`
1323 |     builtin # offset_of($Container, $($fields).+)
     |
    ::: library/core/tests/mem.rs:464:16
     |
     |
464  |     assert_eq!(offset_of!(Beta, x), 0);
     |
     |
     = help: the trait `Sized` is not implemented for `(dyn mem::offset_of_dst::Trait + 'static)`

error[E0277]: the size for values of type `(dyn mem::offset_of_dst::Trait + 'static)` cannot be known at compilation time
     |
     |
1322 | pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
     | ------------------- in this expansion of `offset_of!`
1323 |     builtin # offset_of($Container, $($fields).+)
     |
    ::: library/core/tests/mem.rs:465:16
     |
     |
465  |     assert_eq!(offset_of!(Beta, y), 2);
     |
     |
     = help: the trait `Sized` is not implemented for `(dyn mem::offset_of_dst::Trait + 'static)`
error[E0277]: the size for values of type `mem::offset_of_dst::Extern` cannot be known at compilation time
    --> /checkout/library/core/src/mem/mod.rs:1323:5
     |
     |
1322 | pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
     | ------------------- in this expansion of `offset_of!`
1323 |     builtin # offset_of($Container, $($fields).+)
     |
    ::: library/core/tests/mem.rs:467:16
     |
     |
467  |     assert_eq!(offset_of!(Gamma, x), 0);
     |
     = help: the trait `Sized` is not implemented for `mem::offset_of_dst::Extern`

error[E0277]: the size for values of type `mem::offset_of_dst::Extern` cannot be known at compilation time
error[E0277]: the size for values of type `mem::offset_of_dst::Extern` cannot be known at compilation time
    --> /checkout/library/core/src/mem/mod.rs:1323:5
     |
1322 | pub macro offset_of($Container:ty, $($fields:tt).+ $(,)?) {
     | ------------------- in this expansion of `offset_of!`
1323 |     builtin # offset_of($Container, $($fields).+)
     |
    ::: library/core/tests/mem.rs:468:16
     |
     |
468  |     assert_eq!(offset_of!(Gamma, y), 2);
     |
     = help: the trait `Sized` is not implemented for `mem::offset_of_dst::Extern`

For more information about this error, try `rustc --explain E0277`.
```

</details>

--- a/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/rvalue.rs
@@ -668,11 +668,16 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
 
             mir::Rvalue::NullaryOp(ref null_op, ty) => {
                 let ty = self.monomorphize(ty);
-                assert!(bx.cx().type_is_sized(ty));
                 let layout = bx.cx().layout_of(ty);
                 let val = match null_op {
-                    mir::NullOp::SizeOf => layout.size.bytes(),
-                    mir::NullOp::AlignOf => layout.align.abi.bytes(),
+                    mir::NullOp::SizeOf => {
+                        assert!(bx.cx().type_is_sized(ty));
+                        layout.size.bytes()
+                    }
+                    mir::NullOp::AlignOf => {
+                        assert!(bx.cx().type_is_sized(ty));
+                        layout.align.abi.bytes()
+                    }
                     mir::NullOp::OffsetOf(fields) => {
                         layout.offset_of_subfield(bx.cx(), fields.iter().map(|f| f.index())).bytes()
                     }

--- a/tests/ui/offset-of/offset-of-unsized.rs
+++ b/tests/ui/offset-of/offset-of-unsized.rs
@@ -1,0 +1,15 @@
+// build-pass
+// regression test for #112051
+
+#![feature(offset_of)]
+
+struct S<T: ?Sized> {
+    a: u64,
+    b: T,
+}
+trait Tr {}
+
+fn main() {
+    let _a = core::mem::offset_of!(S<dyn Tr>, a);
+    let _b = core::mem::offset_of!((u64, dyn Tr), 0);
+}

